src/backend/columnar/columnar_tableam.c

```diff
@@ -2104,31 +2164,116 @@ ColumnarProcessUtility(PlannedStmt *pstmt,
 
 	Node *parsetree = pstmt->utilityStmt;
 
-	if (IsA(parsetree, IndexStmt))
+	RangeVar *columnarRangeVar = NULL;
+	List *columnarOptions = NIL;
+
+	switch (nodeTag(parsetree))
 	{
-		IndexStmt *indexStmt = (IndexStmt *) parsetree;
+		case T_IndexStmt:
+		{
+			IndexStmt *indexStmt = (IndexStmt *) parsetree;
 
-		Relation rel = relation_openrv(indexStmt->relation,
-									   indexStmt->concurrent ? ShareUpdateExclusiveLock :
-									   ShareLock);
+			Relation rel = relation_openrv(indexStmt->relation,
+										   indexStmt->concurrent ?
+										   ShareUpdateExclusiveLock :
+										   ShareLock);
 
-		if (rel->rd_tableam == GetColumnarTableAmRoutine())
+			if (rel->rd_tableam == GetColumnarTableAmRoutine())
+			{
+				CheckCitusColumnarVersion(ERROR);
+				if (!ColumnarSupportsIndexAM(indexStmt->accessMethod))
+				{
+					ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+									errmsg("unsupported access method for the "
+										   "index on columnar table %s",
+										   RelationGetRelationName(rel))));
+				}
+			}
+
+			RelationClose(rel);
+			break;
+		}
+
+		case T_CreateStmt:
 		{
-			CheckCitusColumnarVersion(ERROR);
-			if (!ColumnarSupportsIndexAM(indexStmt->accessMethod))
+			CreateStmt *createStmt = castNode(CreateStmt, parsetree);
+			bool no_op = false;
+
+			if (createStmt->if_not_exists)
 			{
-				ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-								errmsg("unsupported access method for the "
-									   "index on columnar table %s",
-									   RelationGetRelationName(rel))));
+				Oid existing_relid;
+
+				/* use same check as transformCreateStmt */
+				(void) RangeVarGetAndCheckCreationNamespace(
+					createStmt->relation, AccessShareLock, &existing_relid);
+
+				no_op = OidIsValid(existing_relid);
 			}
+
+			if (!no_op && createStmt->accessMethod != NULL &&
+				!strcmp(createStmt->accessMethod, COLUMNAR_AM_NAME))
+			{
+				columnarRangeVar = createStmt->relation;
+				createStmt->options = ExtractColumnarRelOptions(createStmt->options,
+																&columnarOptions);
+			}
+			break;
 		}
 
-		RelationClose(rel);
+		case T_CreateTableAsStmt:
+		{
+			CreateTableAsStmt *createTableAsStmt = castNode(CreateTableAsStmt, parsetree);
+			IntoClause *into = createTableAsStmt->into;
+			bool no_op = false;
+
+			if (createTableAsStmt->if_not_exists)
+			{
+				Oid existing_relid;
+
+				/* use same check as transformCreateStmt */
+				(void) RangeVarGetAndCheckCreationNamespace(
+					into->rel, AccessShareLock, &existing_relid);
+
+				no_op = OidIsValid(existing_relid);
+			}
+
+			if (!no_op && into->accessMethod != NULL &&
+				!strcmp(into->accessMethod, COLUMNAR_AM_NAME))
+			{
+				columnarRangeVar = into->rel;
+				into->options = ExtractColumnarRelOptions(into->options,
+														  &columnarOptions);
+			}
+			break;
+		}
```

Comment:
When creating a columnar table from `CREATE TABLE ... AS ...` I notice that the query still gets executed, even if one of the columnar options cause an error. Not a huge problem, I just wonder if it would make sense to throw errors before dispatching to the original utility hook.

```
[local] citus@citus:9700-11392=# CREATE TABLE c3 USING columnar WITH (columnar.chunk_group_row_limit = 1) AS SELECT * FROM t1;
NOTICE:  issuing SELECT a, b, c, d FROM citus.t1_102016 t1 WHERE true
DETAIL:  on server citus@localhost:9701 connectionId: 5
NOTICE:  issuing SELECT a, b, c, d FROM citus.t1_102017 t1 WHERE true
DETAIL:  on server citus@localhost:9702 connectionId: 6
NOTICE:  issuing SELECT a, b, c, d FROM citus.t1_102018 t1 WHERE true
DETAIL:  on server citus@localhost:9701 connectionId: 5
NOTICE:  issuing SELECT a, b, c, d FROM citus.t1_102019 t1 WHERE true
DETAIL:  on server citus@localhost:9702 connectionId: 6
ERROR:  chunk group row count limit out of range
HINT:  chunk group row count limit must be between 1000 and 100000
Time: 44.518 ms
```

I guess this is due to the error being thrown from `SetColumnarRelOptions`, and I think most of the time it is ok to throw the error late, as the work done is not much.

However, `CREATE TABLE ... AS ...` and, in pg15 changing the table access methods could be somewhat heavy IO operations.

Would it make sense to have a precheck of the columnar options before we dispatch `PrevProcessUtilityHook_compat`?

Reply:
Good catch!

Fixed by refactoring. Worth it because it seems like a normal use case. Though with #5927, we have another related issue to fix.